crates/hyperqueue/src/stream/reader/streamdir.rs

```diff
@@ -0,0 +1,464 @@
+    }
+
+    pub fn instance_mut(&mut self, instance_id: InstanceId) -> Option<&mut InstanceInfo> {
+        self.instances
+            .iter_mut()
+            .find(|info| info.instance_id == instance_id)
+    }
+
+    pub fn instance(&self, instance_id: InstanceId) -> Option<&InstanceInfo> {
+        self.instances
+            .iter()
+            .find(|info| info.instance_id == instance_id)
+    }
+}
+
+pub struct StreamDir {
+    paths: Vec<PathBuf>,
+    index: BTreeMap<JobId, BTreeMap<JobTaskId, TaskInfo>>,
+    cache: LruCache<usize, BufReader<File>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Summary {
+    pub n_files: u32,
+    pub n_jobs: u32,
+    pub n_tasks: u64,
+    pub n_streams: u64,
+    pub n_opened: u64,
+    pub stdout_size: u64,
+    pub stderr_size: u64,
+    pub n_superseded: u64,
+    pub superseded_stdout_size: u64,
+    pub superseded_stderr_size: u64,
+}
+
+impl StreamDir {
+    pub fn open(path: &Path, server_uid: Option<&str>) -> crate::Result<Self> {
+        log::debug!("Reading stream dir {}", path.display());
+        let mut paths = Vec::new();
+        let mut server_uids: Set<String> = Set::new();
+        let mut found = false;
+        for entry in std::fs::read_dir(path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path
+                .extension()
+                .and_then(|e| e.to_str())
+                .map(|s| s == "hqs")
+                .unwrap_or(false)
+            {
+                found = true;
+                log::debug!("Discovered {}", path.display());
+                let mut file = BufReader::new(File::open(&path)?);
+                let header = match StreamDir::check_header(&mut file) {
+                    Ok(header) => header,
+                    Err(e) => {
+                        log::debug!(
+                            "Skipping file {}, because reading header fails: {}",
+                            path.display(),
+                            e
+                        );
+                        continue;
+                    }
+                };
+                if let Some(uid) = server_uid {
+                    if uid != header.server_uid.as_str() {
+                        log::debug!("{} ignored because different server uid", path.display());
+                        continue;
+                    }
+                }
+                server_uids.insert(header.server_uid.into_owned());
+                paths.push(path);
+            }
+        }
+
+        if !found {
+            return Err(HqError::GenericError("No log files found".to_string()));
+        }
+
+        if server_uids.len() > 1 {
+            use std::fmt::Write;
+            let mut msg = "Found streams from multiple server instances:\n".to_string();
+            for server_uid in server_uids {
+                writeln!(msg, "{server_uid}").unwrap();
+            }
+            msg.push_str("Use --server-uid=... to specify server uid");
```

Comment:
This might be a bit annoying for users. I assume that in most cases, they will want "read the files for the latest server UID". As a future improvement, we could provide this functionality somehow.

Reply:
It is not clear what is the last server uid. Moreover servers may run in parallel and write to the same folder.